### PR TITLE
TRL SFTTrainer Examples

### PR DESCRIPTION
**Pull request by Satrat**
* Simplified the SparseML `Trainer` to be a barebones class definition, everything is handled by `SessionManagerMixIn`. Removed a bunch of old code from loading recipes, as this is handled by `SparseAutoModelForCausalLM`
* Added new `SFTTrainer` class which adds our mix-in to trl's `SFTTrainer`. The only added code here is to add support for passing in a tokenized dataset to `SFTTrainer`
* Added examples of using `SFTTrainer` for sparse finetuning, both with out dataset preprocessing and TRL's dataset preprocessing

Asana ticket: https://app.asana.com/0/1201735099598270/1206486351032763/f 

### Testing 
See examples in `integrations/huggingface-transformers/tutorials/text-generation/trl_mixin`

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks Sara - this looks really nice

Are there any other features we should flex? I am thinking we might want to look at:
- FSDP
- Distillation

File: src/sparseml/transformers/finetune/session_mixin.py [hunk 1]
```diff
@@ -107,14 +105,17 @@ def __init__(
         self.criterion = torch.nn.CrossEntropyLoss()
 
         model_signature = inspect.signature(self.model.forward)
-        self._model_signature_columns = list(model_signature.parameters.keys())
+        self._signature_columns = list(model_signature.parameters.keys())
```
Comment (bfineran) [inline src/sparseml/transformers/finetune/session_mixin.py:108]: is this due to transformers version?

File: src/sparseml/transformers/finetune/sft_trainer.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# limitations under the License.
+
+from sparseml.transformers.finetune.session_mixin import SessionManagerMixIn
+from trl import SFTTrainer as TRLSFTTrainer
+
+
+__all__ = ["SFTTrainer"]
+
+
+class SFTTrainer(SessionManagerMixIn, TRLSFTTrainer):
```
Comment (bfineran) [inline src/sparseml/transformers/finetune/sft_trainer.py:22]: excellant!

**Review by bfineran — COMMENTED** [review]
looks great overall - should probably move the examples out of src and add a brief readme to go along with it. Debatable whether or not we'd want SFTTrainer out of src as well

File: src/sparseml/transformers/finetune/session_mixin.py (hunk 1 above)
Comment (Satrat) [reply src/sparseml/transformers/finetune/session_mixin.py:108]: Yep, not our recent upgrade though. I had kept this name in from the old trainer code but realized it got renamed a few versions ago

File: src/sparseml/transformers/finetune/examples/test_trl_sft_data.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+response_template = "Answer:"
+collator = DataCollatorForCompletionOnlyLM(response_template, tokenizer=tokenizer)
+
+training_args = TrainingArguments(
+    output_dir=output_dir,
+    num_train_epochs=0.6,
+    logging_steps=50,
+    gradient_checkpointing=True,
+)
```
[the remark this answers is not in the document]
Comment (Satrat) [reply src/sparseml/transformers/finetune/examples/test_trl_sft_data.py:65]: A few things won't work if the native transformers TrainingArguments is used: no support for recipe overrides, no compressed save, no multistage training runs. The mix-in uses these params, so if we wanted to support the non-sparseml TrainingArguments we would have to check each time we reference them. I don't think its worth the extra lines personally

File: src/sparseml/transformers/finetune/sft_trainer.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    def _prepare_dataset(self, dataset, *args, **kwargs):
+        if "input_ids" in dataset.column_names:
+            # dataset is already tokenized, skip preprocessing
+            return dataset
+
+        return super()._prepare_dataset(dataset, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (Satrat) [reply src/sparseml/transformers/finetune/sft_trainer.py:28]: Yes good point, I'll move this to examples. I don't think it makes sense to put the _prepare_dataset override in the MixIn though, its very specific to TRL and was just a way to get around them not having direct support for passing in tokenized datasets

File: src/sparseml/transformers/finetune/examples/test_trl_sft_data.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    tokenizer=tokenizer,
+    recipe=recipe,
+    train_dataset=dataset,
+    formatting_func=formatting_prompts_func,
+    data_collator=collator,
+    args=training_args,
+    max_seq_length=512,
+)
+trainer.train()
+trainer.save_model(output_dir=trainer.args.output_dir)
```
[the remark this answers is not in the document]
Comment (Satrat) [reply src/sparseml/transformers/finetune/examples/test_trl_sft_data.py:78]: yep!

**Review by bfineran — APPROVED** [review]

**Pull request merged**